### PR TITLE
Improvements to MPI_Test and friends

### PR DESCRIPTION
**Pull request by devreal**
This PR is an attempt to improve MPI_Test and friends after I started looking through the code in the context of an application using `MPI_Testsome`. The main issue I'm trying to address is that the test functions invoke the progress engine and return immediately, without rechecking the requests. The typical use-case for `MPI_Test*` is to hide communication latencies by performing other operations until the requests are complete. By not checking whether a call to `opal_progress` changed the state of requests we're forcing applications into another round-trip, potentially adding latency until they can react to completed operations. And then they will need to be checked anyway...

The changes are as follows;

1) Have `opal_progress()` return the number of events it counted from the invocation of the progress callbacks. This is helpful for upper layers to estimate whether the state of some operations *may* have changed.
2) `MPI_Test`: only recheck the request if `opal_progress()` reported an event happened, no need to recheck otherwise.
3) `MPI_Testall`: when encountering an incomplete request, invoke `opal_progress()` immediately and (if `opal_progress()` signaled some completion) continue walking the list of requests. If another incomplete request is found we can bail out immediately, without running until the end of the list.
4) `MPI_Testsome`: similar to `MPI_Testall`, invoke `opal_progress()` when encountering the first incomplete request but continue walking the list of requests, collecting all requests that are complete. Current behavior of `MPI_Testsome` is to not trigger progress if there is at least one complete request, potentially delaying the progress of other operations (nonblocking collectives for example, which require calls into the progress engine).
5) `MPI_Testany`: call into `opal_progress` if no complete request was found and recheck the list if a state change might have occurred. If a complete request is found, progress is not triggered. I'm not sure whether this is a good idea or not but without traversing the full list of request we won't know whether or not progress is necessary. 

I split the changes to each function into individual commits for easier review. In addition to the changes above, I also tried to avoid redundant checks such as checking for a failed request if the request was found to be complete or inactive.

### CODE REVIEW

File: ompi/request/req_test.c
Comment (abouteiller) [inline ompi/request/req_test.c:357]: I think you need to reset num_requests_null_inactive and num_requests_done to 0 when taking the goto

File: ompi/request/req_test.c
Comment (abouteiller) [inline ompi/request/req_test.c:215]: I think you need to. reset num_completed to 0 here as well

**Review by abouteiller — CHANGES_REQUESTED** [review]
Found some potential miscounting in the 'recheck' goto case

File: ompi/request/req_test.c
Comment (bosilca) [reply ompi/request/req_test.c:215]: No need to reset, the code stays in the same loop. The PR text describes more clearly the mechanism, the loop is only done once, the first time it encounters an active, non-completed request calls opal_progress, and then if anything new happened it continues the loop starting from the current element.

File: ompi/request/req_test.c
Comment (bosilca) [inline ompi/request/req_test.c:246]: I think you could have made the code more readable, by only increasing the loop index when the request is completed or in active. This way you would not need the label and goto, and the logic of redoing the same loop index after the opal_progress would have been more natural.

Also adding a continue in the case when you increase the loop index would have removed all the "else" in front of the remaining branches.

File: ompi/request/req_test.c
Comment (bosilca) [inline ompi/request/req_test.c:380]: Same comment as for test_all above.

**Review by abouteiller — APPROVED** [review]

File: ompi/request/req_test.c
Comment (devreal) [reply ompi/request/req_test.c:378]: In this case I prefer having the else branches instead of continue, it makes the code more readable (and the generated assembly should be the same). 

I have removed the use of goto in `MPI_Testsome` and `MPI_Testall`.

File: ompi/request/req_test.c
Comment (bosilca) [inline ompi/request/req_test.c:324]: ?

Comment (devreal) [reply ompi/request/req_test.c:324]: That was a leftover from using goto earlier, fixed.

**Comment by awlauria** [comment]
bot:aws:retest

**Comment by awlauria** [comment]
Any more additional comments/changes needed?

**Comment by devreal** [comment]
I will make some changes, marked it as WIP just now.

**Comment by devreal** [comment]
After some discussions with @bosilca and @abouteiller I have removed the changes to `MPI_Testsome` and `MPI_Testany` because it is not clear whether the rechecking is beneficial under all circumstances. The PR still changes the behavior of `MPI_Testall` in that it short-cuts if the call to `opal_progress` did not signal that anything has been progressed. It also changes `MPI_Test` to only recheck if progress has been signaled.

As a minor optimization in `MPI_Testsome` I have adapted the code such that the request is only tested for the failed state if it is neither null or complete. There is no need to test for the failed state if the request is complete.

**Comment by awlauria** [comment]
Is this good to go since your latest updates? Based on your comments it sounds like it, but just want to make sure.

**Comment by devreal** [comment]
@awlauria Thanks for the reminder. I'd like to have another review by @bosilca or @abouteiller  before merging.

File: ompi/request/req_test.c
Comment (bosilca) [inline ompi/request/req_test.c:232]: fix the comment please.

**Review by bosilca — APPROVED** [review]

**Review by abouteiller — APPROVED** [review]

**Comment by awlauria** [comment]
Please chery-pick to v5.0.x.

Thanks!

**Pull request merged**
